src/host/directio.cpp

```diff
@@ -712,11 +638,20 @@ CATCH_RETURN();
 
     try
     {
-        const auto& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
+        auto& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
+        const auto io = gci.GetVtIoForBuffer(&context);
+        const auto corkLock = io ? io->Cork() : Microsoft::Console::VirtualTerminal::VtIo::CorkLock{};
+
         const auto codepage = gci.OutputCP;
         LOG_IF_FAILED(_ConvertCellsToWInplace(codepage, buffer, requestRectangle));
 
-        RETURN_IF_FAILED(_WriteConsoleOutputWImplHelper(context, buffer, requestRectangle, writtenRectangle));
+        RETURN_IF_FAILED(WriteConsoleOutputWImplHelper(context, buffer, requestRectangle.Width(), requestRectangle, writtenRectangle));
+
+        if (io && io->BufferHasContent())
+        {
+            io->WriteCUP(context.GetTextBuffer().GetCursor().GetPosition());
+            io->WriteAttributes(context.GetAttributes().GetLegacyAttributes());

```

Comment:
Following on from my comments about `WriteAttributes` taking a `TextAttribute` parameter, this is a perfect example of where that would be more appropriate. We've already got a `TexAttribute` instance with the correct VT colors, and we're now mapping it back to a legacy attribute before trying to convert to VT again in the `FormatAttributes` method. That can't be right. Won't we be losing information here? Things like RGB colors, italics, etc.

Reply:
> That can't be right. Won't we be losing information here? Things like RGB colors, italics, etc.

I think it's fine to lose some information here. I'd like to make the console APIs work as well as we can possibly make them work, but I think it's somewhat unreasonable to expect that they mix losslessly with VT output. If someone set RGB colors or italics (or similar) I think it's reasonable that the information got lost after a call to `FillConsoleOutput` or similar.

To me it's similar to how `SetConsoleTextAttribute` surprisingly requires us to send a SGR 0 nowadays, because applications like vim (for Windows) have come to expect that it resets all attributes set via VT.

If you feel differently about this, please let me know!

Reply:
It's hard to say, because it's unlikely to effect me personally, but I'm almost sure this will be a breaking change for someone somewhere, and I can foresee a lot of moaning when that happens!

We could always ship it and wait for someone to complain, but I honestly thought this would be a relatively straightforward fix. It's possible I'm missing the bigger picture though.

Again I'm happy to let you guys decide.

Reply:
Any "native" terminal on Windows must support `GetConsoleScreenBufferInfoEx` and its `wAttributes`. My idea was that if this is the bare minimum any terminal must support, then we can use this as the baseline on which we build the console -> VT translation with the highest possible compatibility to all third-party terminals.

One could argue that we could simply ask for extended attributes, but how would we ask for this information to be exposed? As a SGR string? We can't make assumptions about the attributes of the future after all so it can't be structured data. That would mean terminal authors would have to expose the current attributes in more than one format. Ideally I'd prefer if it remained as easy as possible for terminal authors to implement the "read-back" functionality that conhost requires.

As such I'd like to (very very anxiously) wait and see if anyone is affected by this. If no one gets affected by this it nets us an opportunity to make life for terminal authors simpler.

I'll probably thoroughly test `vim` with RGB attributes, etc., before this ships in 1.22, because I also don't want to anger anyone. 🙈

Reply:
Dustin brought up the excellent point that we could use XTPUSHSGR to properly restore the on terminals that support it. (We could alternatively use DECSC to get broader support.)

I'll try to do that in a follow-up PR! It requires changing the `BufferHasContent()` logic, which I'd prefer doing at a later time.

Reply:
> We could alternatively use DECSC

I was just coming here to suggest that! And I realise now why you wanted to avoid using `TextAttributes`. With this case out of the way, almost everything else you're dealing with is a simple legacy attribute. The only exceptions I think are the places relying on `ReadConsoleOutput*`. And if we can handle the scroll operations with VT, and potentially also the attribute fill, then it's really only the `WriteCharacter`/`FillCharacter` operations that'll depend on the underlying buffer attributes, and that seems really unlikely to be something that anyone would expect to work with RGB colors and VT attributes.

TLDR: I'm a lot happier with where this is going now that I understand it better!

Reply:
I found a flaw with my `Cork` struct: When an exception was thrown, the `Cork` destructor would call `VtIo::_flushNow` and emit partially completed VT sequences. To solve that I had to do a larger change in [`f0db29d`](https://github.com/microsoft/terminal/pull/17510/commits/f0db29ddd1493b30dc5cb13b60e58b25e2e15f3b).

The good news is that this allowed me to add support for DECSC/DECRC.